File: BetterTridents/src/main/java/project/kompass/btk/listener/TridentLootingListener.java
package project.kompass.btk.listener;

import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Mob;
import org.bukkit.entity.Player;
import org.bukkit.entity.Trident;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.loot.LootContext;
import org.bukkit.loot.LootTable;

import java.util.Random;

public class TridentLootingListener implements Listener {

    @EventHandler
    public void onEntityDeath(EntityDeathEvent event) {
        if (!(event.getEntity().getLastDamageCause() instanceof EntityDamageByEntityEvent damageEvent) ||
                !(damageEvent.getDamager() instanceof Trident trident)) {
            return;
        }

        int level = trident.getItemStack().getEnchantmentLevel(Enchantment.LOOTING);
        if (level <= 0 || !(trident.getShooter() instanceof Player shooter)) return;

        LootTable lootTable = ((Mob) event.getEntity()).getLootTable();
        if (lootTable != null) {
            event.getDrops().clear();
            event.getDrops().addAll(lootTable.populateLoot(
                    new Random(),
                    new LootContext.Builder(event.getEntity().getLocation())
                            .lootedEntity(event.getEntity())
                            .killer(shooter)
                            .luck(level)
                            .build()
            ));
        }
    }
}

File: BetterTridents/src/main/kotlin/project/kompass/btk/listener/TridentLootingListener.kt
package project.kompass.btk.listener

import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Mob
import org.bukkit.entity.Player
import org.bukkit.entity.Trident
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityDeathEvent
import org.bukkit.loot.LootContext
import java.util.Random

class TridentLootingListener : Listener {

    @EventHandler
    fun onEntityDeath(event: EntityDeathEvent) {
        val damageEvent = event.entity.lastDamageCause as? EntityDamageByEntityEvent ?: return
        val trident = damageEvent.damager as? Trident ?: return

        val level = trident.itemStack.getEnchantmentLevel(Enchantment.LOOTING)
        if (level <= 0) return
        val shooter = trident.shooter as? Player ?: return

        val mob = event.entity as? Mob ?: return
        val lootTable = mob.lootTable
        if (lootTable != null) {
            event.drops.clear()
            event.drops.addAll(lootTable.populateLoot(
                Random(),
                LootContext.Builder(event.entity.location)
                    .lootedEntity(event.entity)
                    .killer(shooter)
                    .luck(level.toFloat())
                    .build()
            ))
        }
    }
}